### PR TITLE
Initial Implementation

### PR DESCRIPTION
**Pull request by bkendall**
An initial implementation of Ponos! It's going to be a good time.

I'm working on better usage documentation for the README as I'm beginning to use it and make sure this works well!

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+'use strict';
+
+var Server = require('./lib/server');
+
+module.exports = {
+  server: function (opts) { return new Server(opts); },
+  // Errors:
+  TaskError: require('./lib/errors/task-error'),
+  TaskFatalError: require('./lib/errors/task-fatal-error'),
+  TaskMaxRetriesError: require('./lib/errors/task-max-retries-error')
```
Comment (rsandor) [inline index.js]: Is there any reason to expose this on `index`? Isn't it only used internally by the `Worker` class?

File: index.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+'use strict';
+
+var Server = require('./lib/server');
+
+module.exports = {
+  server: function (opts) { return new Server(opts); },
```
Comment (rsandor) [inline index.js]: Should we expose the Server and Worker classes as opposed to this factory method?

Comment (rsandor) [inline index.js]: Also should we expose the logger so they can customize the streams, etc.?

File: lib/errors/task-error.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+ * @class
+ * @param {string} queue Name of the queue that encountered the error.
+ * @param {string} message Message for the task error.
+ * @param {object} [data] Extra data to include with the error, optional.
+ */
+function TaskError (queue, message, data) {
+  Error.call(this);
+  this.setMessageAndData(queue, message, data);
+}
+util.inherits(TaskError, Error);
```
Comment (rsandor) [inline lib/errors/task-error.js]: Should this inherit from `PonosError`?

File: lib/logger.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+// var isObject = require('101/is-object');
+var bunyan = require('bunyan');
+// var Bunyan2Loggly = require('bunyan-loggly').Bunyan2Loggly;
+
+/**
+ * Bunyan logger for ponos.
+ * @author Bryan Kendall
+ * @module ponos:logger
+ */
+module.exports = bunyan.createLogger({
```
Comment (rsandor) [inline lib/logger.js]: Should the logger should be a `Server` instance variable? If we did it that way we could provide accessor and mutators to allow for easy mutation and full swap-out depending on what makes sense in the given implementation.

On the flip side it would mean we'd have to pass it from the server to each of the workers, and the workers to errors... uugghh.

File: lib/server.js [hunk 5]
```diff
@@ -68,64 +67,111 @@ module.exports = Server;
...
- * @return {Promise} A promise that resolves when the server is listening.
+ * Helper function to ensure we have all required task handlers for the queue
+ * array with which the class was defined.
+ * @return {promise} Resolved when all tasks verified.
  */
-Server.prototype.start = function () {
-  var self = this;
-  return this.hermes.connectAsync()
-    .then(Promise.map(self.hermes.queues, self._subscribe));
+Server.prototype._assertHaveAllTasks = function () {
```
Comment (rsandor) [inline lib/server.js]: This is needed at the moment because of how hermes works enforcing a given number of queues up front. In the future we may be able to dispense with this check.

File: lib/worker.js [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+Worker.prototype.run = function () {
+  var baseLogData = {
+    queue: this.queue,
+    job: this.job
+  };
+  return Promise.resolve().bind(this)
+    .then(function checkAttemptCounter () {
+      this.attempt++;
+      // Check if we've exceeded the maximum number of retries
+      if (this.attempt > process.env.WORKER_MAX_RETRIES) {
```
Comment (podviaznikov) [inline lib/worker.js]: Just exploring idea:
do we want to have global WORKER_MAX_RETRIES or it should be task specific?

I don't see any particular case when we would want to have task specific MAX_RETRIES parameter... but maybe you can imagine some interesting cases?

File: lib/server.js [hunk 7]
```diff
@@ -68,64 +67,111 @@ module.exports = Server;
...
+ */
+Server.prototype.setAllTasks = function (map) {
+  return Promise.resolve(Object.keys(map)).bind(this)
+    .each(function (key) { return this.setTask(key, map[key]); });
+};
+
+/**
+ * Assigns a task to a queue.
+ * @param {string} queueName Queue name.
+ * @param {function} task Function to take a job and return a promise.
```
Comment (rsandor) [inline lib/server.js]: Ha nice definition

File: lib/server.js [hunk 8]
```diff
@@ -68,64 +67,111 @@ module.exports = Server;
...
+ * @returns {promise} Resolved when task assigned to queue.
```
Comment (rsandor) [inline lib/server.js]: Resolved when **the** task **is** assigned to **a** queue.

File: lib/worker.js [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    'done'
+  ];
+  fields.forEach(function (f) {
+    if (!exists(opts[f])) {
+      throw new Error(f + ' is required for a Worker');
+    }
+  });
+
+  // manage field defaults
+  fields.push('runNow');
```
Comment (podviaznikov) [inline lib/worker.js]: do we need this line? 
I see that we are setting `runNow` on the line 52 anyway

File: lib/logger.js [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+ * Streams for ponos's bunyan logger.
+ * @return {array} An array of streams for the bunyan logger
+ */
+function getStreams () {
+  var streams = [{
+    level: process.env.LOG_LEVEL_STDOUT,
+    stream: process.stdout
+  }];
+
+  // if (exists(process.env.LOGGLY_TOKEN)) {
```
Comment (rsandor) [inline lib/logger.js]: WIP stuff I assume, going to figure out a good pattern for handling logging while implementing your next project?

File: lib/worker.js [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+  return new Worker(opts);
+};
+
+/**
+ * Runs the worker. If the task for the job fails, then this method will retry
+ * the task (with an exponential backoff) a number of times defined by the
+ * environment of the process.
+ * @returns {promise} Promise resolved once the task succeeds or fails.
+ */
+Worker.prototype.run = function () {
```
Comment (podviaznikov) [inline lib/worker.js]: Would it be easy to add in the future worker timeout? I don't know Promises well so don't know how easy/beautiful it be.

I think we would need to have timeout that should cause either worker to rerun or even nack job?

File: lib/worker.js [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+      throw new Error(f + ' is required for a Worker');
+    }
+  });
+
+  // manage field defaults
+  fields.push('runNow');
+  opts = pick(opts, fields);
+  defaults(opts, {
+    runNow: true,
+    // not necessarily user opts
```
Comment (rsandor) [inline lib/worker.js]: They just aren't

File: lib/worker.js [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+        throw new TaskMaxRetriesError(this.queue, errData);
+      }
+    })
+    .then(function runTheTask () {
+      var attemptData = { attempt: this.attempt };
+      log.info(assign({}, baseLogData, attemptData), 'Running task');
+      return this.task(this.job);
+    })
+    .then(function successDone (result) {
+      log.info(assign({}, baseLogData, { result: result }), 'Job complete');
```
Comment (podviaznikov) [inline lib/worker.js]: seems like you are calling `assign` always with `{}` as first argument.
Can we use `put` for that? I think `put` acts like `merge` and creates new object instead of modifying existing. So it does what you want but with 2 arguments

File: lib/worker.js [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+    })
+    // TODO(bryan): do we want to do something special for TaskError?
+    .catch(function unknownErrRetry () {
+      var attemptData = { nextAttemptDelay: this.retryDelay };
+      log.warn(assign({}, baseLogData, attemptData), 'Task failed, retrying');
+
+      // Try again after a delay
+      return Promise.resolve()
+        .delay(this.retryDelay).bind(this)
+        .then(function () {
```
Comment (podviaznikov) [inline lib/worker.js]: give name to this anonymous function. Seems like you are giving names to all anonymous functions above

File: index.js [hunk 15]
```diff
@@ -0,0 +1,11 @@
...
+  // Errors:
+  TaskError: require('./lib/errors/task-error'),
```
Comment (podviaznikov) [inline index.js]: I like those errors. Nice job :

File: lib/worker.js [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+      log.info(assign({}, baseLogData, attemptData), 'Running task');
+      return this.task(this.job);
+    })
+    .then(function successDone (result) {
+      log.info(assign({}, baseLogData, { result: result }), 'Job complete');
+      return this.done();
+    })
+    // if it's a known type of error, we can't accomplish the task
+    .catch(TaskFatalError, TaskMaxRetriesError, function knownErrDone (err) {
+      error.log(err);
```
Comment (podviaznikov) [inline lib/worker.js]: I assume at this point it's guaranteed that `err` is defined, right?

File: lib/error.js [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+  ErrorCat.apply(this, arguments);
+}
+util.inherits(PonosError, ErrorCat);
+
+/**
+ * Logs errors via default ponos logger.
+ * @param {error} err PonosError to log.
+ * @see error-cat
+ */
+PonosError.prototype.log = function (err) {
```
Comment (rsandor) [inline lib/error.js]: So the way we handle error logging in our internal projects is to use [error-cat](https://github.com/runnable/error-cat). Should we eat our own cat food? We could technically have ponos use error-cat and monitor-dog by default. And give the user a way of redefining the instances (in the case they have done custom implementations in their projects).

Comment (rsandor) [inline lib/error.js]: If we end up going down the use all the things route then we'll definitely not need this.

**Comment by podviaznikov** [comment]
I like implementation. Good job. Have just few minor comments/questions

File: lib/worker.js [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+    .then(function runTheTask () {
+      var attemptData = { attempt: this.attempt };
+      log.info(assign({}, baseLogData, attemptData), 'Running task');
+      return this.task(this.job);
+    })
+    .then(function successDone (result) {
+      log.info(assign({}, baseLogData, { result: result }), 'Job complete');
+      return this.done();
+    })
+    // if it's a known type of error, we can't accomplish the task
```
Comment (rsandor) [inline lib/worker.js]: This is a vague comment, should it be axed?

File: lib/worker.js [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+    .catch(TaskFatalError, TaskMaxRetriesError, function knownErrDone (err) {
+      error.log(err);
+      // If we encounter a fatal error we should no longer try to schedule
+      // the job.
+      return this.done();
+    })
+    // TODO(bryan): do we want to do something special for TaskError?
```
Comment (rsandor) [inline lib/worker.js]: After thinking about it I remembered that I made `TaskError` as an ease-of-implementation tool. Basically it defined a strict form for the types of errors that can be explicitly thrown from a task handler. From that perspective I don't see a reason why we should be handling them explicitly in the run method.

File: lib/worker.js [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+    .catch(function unknownErrRetry () {
+      var attemptData = { nextAttemptDelay: this.retryDelay };
+      log.warn(assign({}, baseLogData, attemptData), 'Task failed, retrying');
```
Comment (rsandor) [inline lib/worker.js]: I don't know how I feel about this style of `assign` logging. I almost feel like you have to do more work to figure out what is going on (by scrolling back up to the top of the function, etc.) than it would take if we simply included all the log fields explicitly on the statement.

Furthermore, it makes it harder to start jamming way too much information into any individual log line (as it becomes immediately apparent at the point of the log entry).

File: package.json [hunk 21]
```diff
@@ -4,7 +4,10 @@
   "description": "An opinionated queue based worker server for node.",
   "main": "index.js",
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "coverage": "istanbul cover ./node_modules/.bin/_mocha -- --recursive --reporter dot test/unit",
+    "coverage-check": "istanbul check-coverage && echo 'Coverage check successful!'",
+    "lint": "eslint .",
+    "test": "mocha --recursive --reporter dot"
```
Comment (rsandor) [inline package.json]: Dots? Really?

Comment (rsandor) [inline package.json]: Also there is a lot of log junk going on when testing, should we explicitly set the log level to none?

File: test/unit/server.js [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+var ponos = require('../../');
+
+var tasks = {
+  'test-queue-01': worker,
+  'test-queue-02': worker
+};
+function resolvePromise () { return Promise.resolve(); }
+function worker (job, done) {}; // eslint-disable-line no-unused-vars
+
+describe('Ponos Server', function () {
```
Comment (rsandor) [inline test/unit/server.js]: Rename: `Server`

File: test/unit/worker.js [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+var sinon = require('sinon');
+
+var TaskFatalError = require('../../lib/errors/task-fatal-error');
+var TaskMaxRetriesError = require('../../lib/errors/task-max-retries-error');
+var Promise = require('bluebird');
+var Worker = require('../../lib/worker');
+var assign = require('101/assign');
+var error = require('../../lib/error');
+
+describe('Ponos Worker', function () {
```
Comment (rsandor) [inline test/unit/worker.js]: Rename: `Worker`

File: test/unit/server.js [hunk 24]
```diff
@@ -0,0 +1,128 @@
...
+var assert = chai.assert;
+var sinon = require('sinon');
+
+var Promise = require('bluebird');
+var Worker = require('../../lib/worker');
+var hermes = require('runnable-hermes');
+var noop = require('101/noop');
+var ponos = require('../../');
+
+var tasks = {
```
Comment (rsandor) [inline test/unit/server.js]: Nit: generally I like to keep local fixtures and methods inside the describe.

File: test/unit/server.js [hunk 25]
```diff
@@ -0,0 +1,128 @@
...
+    sinon.stub(server.hermes, 'connectAsync', resolvePromise);
+    sinon.stub(server.hermes, 'subscribeAsync', resolvePromise);
+  });
+  afterEach(function () {
+    server.hermes.connectAsync.restore();
+    server.hermes.subscribeAsync.restore();
+  });
+  after(function () { Worker.create.restore(); });
+
+  describe('Constructor', function () {
```
Comment (rsandor) [inline test/unit/server.js]: How are you possibly getting full coverage here? You're not testing any of the `RABBITMQ_*` environment variables, nor determining if they actually default correctly. Is it possible that istanbul is not smart enough to catch in-line logical ORs?

File: test/unit/server.js [hunk 26]
```diff
@@ -0,0 +1,128 @@
...
+      return assert.isFulfilled(server.setAllTasks(tasks))
+        .then(function () {
+          assert.equal(server.setTask.callCount, numTasks);
+          assert.equal(Object.keys(server._tasks).length, numTasks);
+        });
+    });
+  });
+
+  describe('start', function () {
+    describe('missing tasks', function () {
```
Comment (rsandor) [inline test/unit/server.js]: Rename: `without tasks`, reads better... `Server start without tasks should fail`

File: lib/error.js (hunk 17 above)
Comment (anandkumarpatel) [inline lib/error.js]: I agree with making this redefine-able, since we would also want to have loggly put out logs here.

File: lib/server.js [hunk 27]
```diff
@@ -68,64 +67,111 @@ module.exports = Server;
...
+  });
 
-  if (!Array.isArray(this.opts.queues)) {
-    throw new Error('ponos.Server: missing required `queues` option.');
-  }
+  if (this.opts.hermes) {
+    this.hermes = this.opts.hermes;
+  } else {
+    if (!Array.isArray(this.opts.queues)) {
+      throw new Error('ponos.Server: missing required `queues` option.');
```
Comment (anandkumarpatel) [inline lib/server.js]: whats your thoughts on using error cat to create the thrown error? so we get this in rollbar.
or do you think we will never hit the thrown case in prod since test *should* catch this?

File: lib/errors/task-error.js [hunk 28]
```diff
@@ -0,0 +1,44 @@
...
+ * Error type for task rejection promises that indicated to the worker that
+ * something went wrong, but that it should attempt the task again. Useful for
+ * handling any sort of issue that may have a temporal component (network
+ * connectivity, etc.)
+ * @class
+ * @param {string} queue Name of the queue that encountered the error.
+ * @param {string} message Message for the task error.
+ * @param {object} [data] Extra data to include with the error, optional.
+ */
+function TaskError (queue, message, data) {
```
Comment (rsandor) [inline lib/errors/task-error.js]: This doesn't seem to have testing.

File: lib/worker.js (hunk 6 above)
Comment (rsandor) [inline lib/worker.js]: Per our conversation yesterday I think it's a good idea to allow tasks to be set with an explicit number of retries. Shouldn't be too hard, would need to change the way `setTask` and `setAllTasks` works to allow the passing options.

Comment (rsandor) [inline lib/worker.js]: e.g.
```
server.setTask('cluster-deprovision', someFn, { retries: 20 });
```

File: lib/worker.js (hunk 16 above)
Comment (rsandor) [inline lib/worker.js]: Yes, it's part of the `.catch` specification, you will always have an `err` here.

**Comment by rsandor** [comment]
Great job with the implementation! Overall it looks really nice, my only concerns are:

1. **Logging, Reporting, Monitoring** - We need to find a good pattern for allowing ponos to have these right out of the box, but also allow the user to redefine them. Hopefully we can find a good way to do this without being too "cludgy" about it.
2. **More Testing** - The coverage metrics (as mentioned above) are suspect, I can see at least four lines that are not actually fully tested, but are reported as such by istanbul. Furthermore, we should definitely add functional tests to ensure we can connect to queues, and perform jobs.

File: lib/errors/task-error.js (hunk 3 above)
Comment (rsandor) [inline lib/errors/task-error.js]: Oh I see, perhaps we should call this `PonosErrorCat`?

File: lib/worker.js (hunk 11 above)
Comment (bkendall) [inline lib/worker.js]: Yes. [`Promise.timeout([ms])`](https://github.com/petkaantonov/bluebird/blob/master/API.md#timeoutint-ms--string-message---promise)

Probably will add timeouts in this initial implementation.

File: lib/worker.js (hunk 13 above)
Comment (bkendall) [inline lib/worker.js]: thanks - I forgot `put` will merge objects...

File: lib/worker.js (hunk 14 above)
Comment (bkendall) [inline lib/worker.js]: :+1:

File: index.js (hunk 2 above)
Comment (bkendall) [inline index.js]: > Should we expose the Server and Worker classes

Why would we need to expose the Worker class?

File: index.js (hunk 1 above)
Comment (bkendall) [inline index.js]: nope. don't need `TaskMaxRetriesError`, specifically...

File: lib/errors/task-error.js (hunk 3 above)
Comment (bkendall) [inline lib/errors/task-error.js]: `PonosError` was renamed `PonosErrorCat`

File: lib/logger.js (hunk 4 above)
Comment (bkendall) [inline lib/logger.js]: yeah, I'm still not entirely sure of what to do with the logger at the moment... maybe a quick chat Monday

File: lib/server.js (hunk 5 above)
Comment (bkendall) [inline lib/server.js]: Yup.

File: lib/worker.js (hunk 9 above)
Comment (bkendall) [inline lib/worker.js]: yes - basically, the `pick` below is making sure `opts` only has the fields we specify, so we just need to push it here (appending it to the required fields)

File: lib/server.js (hunk 27 above)
Comment (bkendall) [inline lib/server.js]: This one is tricky because it's in the constructor. The _service_ that is created using `ponos` will most likely _also_ have logging, and if it's failing to even create the server, I imagine it will catch this 'missing options' error and report it itself (or a health check fails, etc). All that to say, I don't think this should _report_ the error yet, because the server can't even be initialized

File: README.md [hunk 29]
```diff
@@ -1,6 +1,60 @@
...
+
+From a high level, Ponos is used to create a worker server that responds to jobs provided from RabbitMQ. The user defines handlers for each queue's jobs that are invoked by Ponos.
+
+Ponos has built in support for retrying and catching specific errors, which are described below.
+
+### Workers
+
+// TODO(bryan)
+
+### Example
```
Comment (podviaznikov) [inline README.md]: Maybe you can add a section on errors and which error should I use in each situation. Or maybe add an example of one job using all errors

File: lib/errors/task-error.js [hunk 30]
```diff
@@ -0,0 +1,44 @@
+'use strict';
+
+var util = require('util');
+var defaults = require('101/defaults');
+var isObject = require('101/is-object');
+
+/**
+ * Error type for task rejection promises that indicated to the worker that
```
Comment (podviaznikov) [inline lib/errors/task-error.js]: non-native English speaker but not sure about this sentence. Double check that it makes sense to you

File: lib/server.js [hunk 31]
```diff
@@ -68,64 +37,115 @@ module.exports = Server;
...
+ * Helper function to subscribe to all queues.
+ * @return {promise} Resolved when queues are all subscribed.
+ */
+Server.prototype._subscribeAll = function () {
+  return Promise.resolve(this.hermes.queues).bind(this)
+    .map(this._subscribe);
+};
+
+/**
  * Helper method that subscribes to a given queue name.
```
Comment (podviaznikov) [inline lib/server.js]: on line 127 you are using "Helper function" and here "Helper method".

File: lib/worker.js (hunk 11 above)
Comment (podviaznikov) [inline lib/worker.js]: nice. Good to know that it would be easy to add this. Don't care about initial version

File: lib/worker.js (hunk 9 above)
Comment (podviaznikov) [inline lib/worker.js]: why do we need this flag? Can't the user of this module call `run` herself?

**Comment by podviaznikov** [comment]
quick questions: 
1. do you think domains should be used in the client of the module?
2. is using of domains still needed with Promises?

File: lib/worker.js (hunk 16 above)
Comment (podviaznikov) [inline lib/worker.js]: good to know. That what I assumed. Need to read more about Promises

File: lib/worker.js (hunk 9 above)
Comment (bkendall) [inline lib/worker.js]: mostly for testing, so you can create a worker in a before and then run it in the test

**Comment by bkendall** [comment]
@podviaznikov to put a similar answer that we discussed in writing: no, I do not believe domains will still be required. Since starting the server lives in a promise, any errors thrown should be caught and handled by them (unless you have some other situation in mind)

File: lib/logger.js (hunk 4 above)
Comment (bkendall) [inline lib/logger.js]: for now, I think leaving Ponos to do it's own logging to stdout and (optionally) loggly is good. Tasks that are defined by users cano handle any error reporting they need.

File: lib/logger.js [hunk 32]
```diff
@@ -0,0 +1,42 @@
+'use strict';
+
+var exists = require('101/exists');
+var bunyan = require('bunyan');
+var Bunyan2Loggly = require('bunyan-loggly').Bunyan2Loggly;
```
Comment (anandkumarpatel) [inline lib/logger.js]: this had memory leaks, we are removing bunyan2loggly in favor of just writing to stdout.

File: lib/logger.js [hunk 33]
```diff
@@ -0,0 +1,42 @@
...
+function getStreams () {
+  var streams = [{
+    level: process.env.LOG_LEVEL_STDOUT,
+    stream: process.stdout
+  }];
+
+  if (exists(process.env.LOGGLY_TOKEN)) {
+    streams.push({
+      level: process.env.LOG_LEVEL_STDOUT,
+      stream: new Bunyan2Loggly({
```
Comment (anandkumarpatel) [inline lib/logger.js]: :boom:

File: lib/server.js (hunk 27 above)
Comment (anandkumarpatel) [inline lib/server.js]: as per discussion downstairs, jobs will handle errors so I am fine with this.

File: package.json [hunk 34]
```diff
@@ -26,7 +31,19 @@
   "dependencies": {
     "101": "^1.0.0",
     "bluebird": "^2.9.34",
+    "bunyan": "^1.5.1",
+    "bunyan-loggly": "0.0.5",
```
Comment (anandkumarpatel) [inline package.json]: :boom:

File: README.md [hunk 35]
```diff
@@ -1,6 +1,122 @@
...
+  queues: Object.keys(tasks);
+});
+
+// Set tasks for workers handling jobs on each queue
+server.setAllTasks(tasks);
+
+// Start the server!
+server.start()
+  .then(function () { console.log('Server started!'); })
+  .catch(function (err) { console.error('Server failed', err); })
```
Comment (rsandor) [inline README.md]: Missing ;

File: test/unit/error.js [hunk 36]
```diff
@@ -0,0 +1,22 @@
...
+
+var chai = require('chai');
+chai.use(require('chai-as-promised'));
+var assert = chai.assert;
+var sinon = require('sinon');
+
+var error = require('../../lib/error');
+var log = require('../../lib/logger');
+
+describe('PonosError', function () {
```
Comment (rsandor) [inline test/unit/error.js]: PonosErrorCat as opposed to PonosError

File: test/unit/server.js (hunk 25 above)
Comment (bkendall) [inline test/unit/server.js]: this has been addressed w/ the tests

**Comment by anandkumarpatel** [comment]
LGTM
